http/README.md

```diff
@@ -0,0 +1,159 @@
+    1. require authentication to (some) paths below `.well-known/libp2p`, and to enforce ACLs
+    2. stealth mode: return 404 for paths below `.well-known/libp2p`, *unless* the client has already authenticated itself, thereby hiding the fact that it runs a libp2p server, even if probed explicitly
+
+## Certificates
+
+libp2p doesn’t prescribe how nodes obtain the TLS certificate to secure the HTTPS connection. Since browsers are expected to connect to the node, the certificate’s trust chain must end in the browser’s trust store.
+
+This is somewhat tricky in a p2p context, as nodes might not have a (sub)domain, which for many CAs is a requirement to obtain a certificate. Specifically, Let’s Encrypt doesn’t support IP certificates at the moment. ZeroSSL does, however, this requires setting up a (free) account.
+
+To speed of server authentication, a node MAY include the libp2p TLS extension in its certificate. Note that this is currently not possible when using Let’s Encrypt, since the libp2p TLS extension is not whitelisted by LE. Not every HTTP client will have access to the TLS certificate (for example, browsers usually don’t expose an API for that), but if an HTTP client does, it SHOULD use that information.
+
+## Authentication
+
+Traditionally, libp2p was built on the assumption that both peers authenticate each other during the libp2p handshake. libp2p+HTTP acknowledges that this isn’t always possible, or even desirable, and that different use cases call for different authentication modes. For example, a server might offer a certain set of services to any client, like a HTTP webserver does.
+
+### Server Authentication
+
+Since HTTP requests are independent from each other (they are not bound to a single connection, and when using HTTP/1.1, will actually use different connections), the server needs to authenticate itself on every single request.
+
+As browsers don’t expose an API to access details of the TLS certificate used, nor allow any access to the (an exporter to) the TLS master secret, server authentication is a bit more contrived than one might initially expect.
+
+To request the server to authenticate, the client sets the `libp2p-server-auth` HTTP header to a randomly generated ASCII string of at least 10 (and a maximum of 100) characters. The server signs the following string using its host key:
+
+```
+"libp2p-server-auth:" || the value of the libp2p-server-auth header || "libp2p-server-domain:" || the domain (including subdomains)
+```
+
+It then sets the following two HTTP headers on the response:
+
+1. `libp2p-server-pubkey`: its public key (from the libp2p key pair)
+2. `libp2p-server-auth-signature`: the signature derived as described above
+
+When requesting server authentication, the client MUST check that these two header fields are present, and MUST check the signature. It MUST NOT process the response if either one of these checks fails
+
+### Client Authentication
```

Comment:
Can we use https://webauthn.guide/#about-webauthn here instead of making our own protocol? I'd rest easier if we leveraged existing work here. I need to think about tradeoffs some more (will it be annoying outside of the browser context?), but generally want to avoid custom auth schemes if we can.

Reply:
That would be fantastic.

Reply:
We still can, but that will be out of scope for this spec. This spec will outline *a* way of doing auth, not *the* way. Other ways will be supported.